src/reflect/scala/reflect/internal/util/SourceFile.scala

```diff
@@ -172,11 +172,32 @@ class BatchSourceFile(val file : AbstractFile, content0: Array[Char]) extends So
 
   private lazy val lineIndices: Array[Int] = {
     def calculateLineIndices(cs: Array[Char]) = {
-      val buf = new ArrayBuffer[Int]
-      buf += 0
-      for (i <- 0 until cs.length) if (isAtEndOfLine(i)) buf += i + 1
-      buf += cs.length // sentinel, so that findLine below works smoother
-      buf.toArray
+      var len = 2 // we add 0 at start and cs.length at end
+      // count how many characters are `isAtEndOfLine`
+      val countEOLs: Int = {
+        var i = 0
+        var c = 0
+        while (i < cs.length){
+          if (isAtEndOfLine(i))
+            c += 1
+          i += 1
+        }
+        c
+      }
+      val res = new Array[Int](countEOLs + 2)
+      res(0) = 0
+      res(res.length - 1) = cs.length
+      // fill res with positions of EOL 
+      var j = 1
+      var i = 0
+      while(i < cs.length){
+        if (isAtEndOfLine(i)) {
+          res(j) = i + 1
+          j += 1
+        }
+        i += 1
+      }
+      res
```

Comment:
It's more natural perhaps to do one pass and increase the `Array[Int]` in the usual way (maybe not double size on reallocation, but bump by an increment). Initial size should be `cs.length / avglen` however you guess avglen. Is that possible in less than 5x LOC? I'd have to look whether the sentinel means the final array can have arbitrary size.

Reply:
Oh, `ArrayBuilder.ofInt`.

Reply:
Although that would still avoid the `Integer` allocations, it would still have the extra allocations of the resizing arrays. The reason why those resizing always double is because it keeps the total allocation within a 2x factor from the final one.

My guess would be that running twice through an array of characters would not take as long as those extra array allocations. But this is a complete guess, not supported by actual benchmarks.

Reply:
BTW one usage does take the array len. In the other PR, I didn't add a counter, so a resize is always required there. If pressed, I'd add the counter and check that the array is not reallocated. Compiler could actually track line lengths and make a good guess for first allocation.